acmwebsite/controllers/survey.py

```diff
@@ -1,34 +1,68 @@
-from tg import expose, redirect, validate, flash, url, lurl, abort, require, request
-from tg.predicates import has_permission, not_anonymous
+from tg import abort, expose, flash, redirect, request, require
+from tg.predicates import has_permission
 
 from acmwebsite.lib.base import BaseController
-from acmwebsite.lib.helpers import log
-from acmwebsite.model import DBSession, Survey, SurveyResponse, SurveyData, User
+from acmwebsite.model import DBSession, Survey, SurveyData, SurveyResponse
 
-def survey_fields(survey):
-    return [{'name': f.name, 'type': f.type} for f in survey.fields]
-
-def response_to_dict(response):
-    out = {'name': response.name, 'email': response.email}
-    for item in response.data:
-        out[item.field.name] = item.field.type_object().from_contents(item.contents)
-    return out
 
 class SurveyController(BaseController):
     def __init__(self, survey):
         self.survey = survey
 
+    @expose('acmwebsite.templates.survey_results')
+    @require(has_permission('admin'))
+    def results(self, number=None, order_by=None, reverse=False):
+        if type(reverse) is str:
+            reverse = reverse == 'True'
+
+        responses = self._survey_responses()
+        if order_by:
+            responses.sort(key=lambda x: x.get(order_by), reverse=reverse)
```

Comment:
Sort on the SQL server.

Reply:
I thought we already fetched the entire `Survey` on line 106...

    survey = DBSession.query(Survey).filter(Survey.id == sid).first()

Does SQLAlchemy not load in the joined tables until they are accessed?

Reply:
Correct. See http://docs.sqlalchemy.org/en/latest/orm/loading_relationships.html

if you want it to be loading the joined tables, use joined load. Although I really don't think that is the solution to the problem (you can encode an order by here... it may just not be obvious).